README.md
# cm4io-fan

kernel module and device tree overlay to add support for the EMC2301 fan controller on the Raspberry Pi Compute Module 4 IO Board.

*Works with 5.10+ 64-bit kernels only.*

Uses Traverse Technologies' EMC2301 [hwmon driver](https://gitlab.traverse.com.au/ls1088firmware/traverse-sensors) for their [ten64](https://www.crowdsupply.com/traverse-technologies/ten64) board, which you should definitely check out because it's awesome.

## Usage
1. Install dkms if you haven't already:
```
sudo apt install dkms
```
2. Download the latest source .tar.gz from the [releases](https://github.com/neg2led/cm4io-fan/releases/) page
3. Untar it to `/usr/src/cm4io-fan-<version>` and run the dkms install:
```
tar -xzvf 0.1.1.tar.gz -C /usr/src/
sudo dkms install cm4io-fan/0.1.1
```
4. Add these lines to your /boot/config.txt (adjust the rpm values for your specific fan, defaults are 3500 / 5500) and reboot.
   See below for more config options.
```
# Enable I2C bus 1 on VideoCore (/dev/i2c-10 in Raspberry Pi OS)
dtparam=i2c_vc=on
# Enable CM4 IO Board fan controller
dtoverlay=cm4io-fan,minrpm=1000,maxrpm=5000
```
5. Some distributions may not automatically load the kernel module despite the devicetree entry; Raspberry Pi OS does, Ubuntu 21.10 does not, etc. 
   To make the module load, edit `/etc/modules` (or make a new file called `/etc/modules-load.d/cm4io-fan.conf`), adding this line:
```
emc2301
```

## Install from git
1. Install dkms if you haven't already:
```
sudo apt install dkms
```
2. Clone the repo
```
mkdir -p ~/src
cd ~/src
git clone https://github.com/neg2led/cm4io-fan.git
cd cm4io-fan
```
3. Run install.sh, feel free to inspect it yourself first. It will archive the current HEAD to /usr/src with an appropriate version, and run DKMS.


## Config options
The device tree overlay has a few options, here's the equivalent of a `/boot/overlays/README` info section:

```
Name:   cm4io-fan
Info:   Raspberry Pi Compute Module 4 IO Board fan controller
Load:   dtoverlay=cm4io-fan,<param>[=<val>]
Params: minrpm              RPM target for the fan when the SoC is below 
                            mintemp (default 3500)
        maxrpm              RPM target for the fan when the SoC is above
                            maxtemp (default 5500)
        midtemp             Temperature (in millicelcius) at which the fan
                            begins to speed up (default 50000)
        midtemp_hyst        Temperature delta (in millicelcius) below mintemp
                            at which the fan will drop to minrpm (default 2000)
        maxtemp             Temperature (in millicelcius) at which the fan 
                            will be held at maxrpm (default 70000)
        maxtemp_hyst        Temperature delta (in millicelcius) below maxtemp
                            at which the fan begins to slow down (default 2000)
```

### many thanks to:
- [taudac/taudac-driver-dkms](https://github.com/taudac/taudac-driver-dkms) for helping me work out how the shit DKMS works
- [Traverse Technologies](https://traverse.com.au) for the [kernel module](https://gitlab.traverse.com.au/ls1088firmware/traverse-sensors)


debian/changelog
cm4io-fan-dkms (0.2.0-0) buster; urgency=medium

  * Fix compatibility issue with Ubuntu 21.10
  * Fix functionality with low-RPM fans

 -- Andrew Powers-Holmes <andrew@omnom.net>  Sun, 01 May 2022 18:30:00 +1000
cm4io-fan-dkms (0.1.0-0) buster; urgency=medium

  * Initial package.

 -- Andrew Powers-Holmes <andrew@omnom.net>  Sat, 08 Apr 2021 22:40:00 +1000


install.sh
#!/usr/bin/env bash
set -e
datestamp=`date -I`

# version/name strings
DRV_NAME="cm4io-fan"
DRV_VERSION="0.2.0"
DRV_DIR="/usr/src/${DRV_NAME}-${DRV_VERSION}"

# check kernel arch
KERN_ARCH=$(uname -m)
# uncomment this to override kernel arch check, I don't think this works on 32-bit, but maybe?
#KERN_ARCH="aarch64"

if [[ ${KERN_ARCH} -ne "aarch64" ]]; then
    echo "Error! aarch64 kernel not detected"
    echo "If you'd like to continue anyway, uncomment the override in install.sh"
    exit 1
fi

# make git archive the repo to the right spot
echo -n "Copying source to ${DRV_DIR}..."
[[ -d ${DRV_DIR} ]] && rm -r ${DRV_DIR}
mkdir -p ${DRV_DIR} && git archive HEAD | tar -x -C "/usr/src/${DRV_NAME}-${DRV_VERSION}"
echo "Done"

# run dkms
echo "Running DKMS install..."
sudo dkms install "${DRV_NAME}/${DRV_VERSION}"


# add line to /boot/config.txt
echo -n "Update config.txt... "
if (grep -q "^dtoverlay=${DRV_NAME}" /boot/config.txt); then
    echo "line already present, no change"
else
    sudo echo -e "dtparam=i2c_vc=on\ndtoverlay=cm4io-fan,minrpm=500,maxrpm=2500" >> /boot/config.txt
    echo "line added, edit to adjust rpm/temp settings."
fi

echo "Done!"
exit 0


emc17xx/emc17xx.c
/*
 * Driver for the Microchip/SMSC EMC1704 Temperature and Voltage monitor
 *
 * Copyright (C) 2018-2019 Traverse Technologies
 * Author: Mathew McBride <matt@traverse.com.au>
 *
 * SPDX-License-Identifier: GPL-2.0
 *
 */

#include <linux/err.h>
#include <linux/hwmon.h>
#include <linux/hwmon-sysfs.h>
#include <linux/i2c.h>
#include <linux/kernel.h>
#include <linux/module.h>

/* High side only for now */
#define EMC1704_INTERNAL_TEMP 0x38
#define EMC1704_INTERNAL_TEMP_LOW 0x39
#define EMC1704_EXTERNAL1_TEMP 0x3A
#define EMC1704_EXTERNAL1_TEMP_LOW 0x3A
#define EMC1704_EXTERNAL2_TEMP 0x3C
#define EMC1704_EXTERNAL2_TEMP_LOW 0x3D
#define EMC1704_EXTERNAL3_TEMP 0x3E
#define EMC1704_EXTERNAL3_TEMP_LOW 0x3F

#define EMC1704_SOURCE_VOLT_REGISTER 0x58
#define EMC1704_SENSE_VOLT_REGISTER 0x55

/* As we can't do floating point, use the 'human' interpretation values given
 * in the EMC1704 datasheet, in uV (1e6).
 * The correct formula is Volts = 23.9883 * (Vsource/4094)
 */
static const int32_t emc1704_source_weights[] = {
	0,0,0,0,0,11700,23400,46900,93800,187500,375000,750000,1500000,3000000,6000000,12000000
};


/* Return the converted value from the given register in uV or mC */
static int emc1704_get_value(struct i2c_client *i2c, u8 reg, int *result)
{
	int mval,tempval,i;

	int8_t highval;
	uint8_t lowval;

	#ifdef DEBUG
	printk(KERN_INFO "Reading register %X\n",reg);
	#endif

	highval = i2c_smbus_read_byte_data(i2c, reg);
	#ifdef DEBUG
	printk(KERN_INFO "Got value %d\n",highval);
	#endif

	lowval = i2c_smbus_read_byte_data(i2c,reg+1);
	#ifdef DEBUG
	printk(KERN_INFO "Got low value %d\n",lowval);
	#endif

	if (reg >= EMC1704_INTERNAL_TEMP && reg <= EMC1704_EXTERNAL3_TEMP) {
		/* The low bits have a weight of .125C */
		lowval = lowval >> 5;
		mval = (highval * 1000) + (lowval * 125);
		#ifdef DEBUG
		printk(KERN_INFO "Calculated value %d\n",mval);
		#endif
		*result = mval;
	} else if (reg == EMC1704_SOURCE_VOLT_REGISTER) {
		tempval = (highval << 8) | lowval;
		mval = 0;
		for(i=15; i>0; i--) {
		if ((tempval & BIT(i)) == BIT(i)) {
			mval += emc1704_source_weights[i];
			#if DEBUG
			printk(KERN_INFO "Bit %d Value now %d\n",i,mval);
			#endif
		}
		*result = mval;
		}
	} else if (reg == EMC1704_SENSE_VOLT_REGISTER) {
		tempval = ((highval & 0x7F) >> 4) | (lowval >> 4);
		/* We masked out the sign bit above */
		if (highval < 0) {
			tempval = 0-tempval;
		}
		*result = tempval;
	} else {
		*result = -0;
		return -EINVAL;
	}

	return 0;
}


static ssize_t emc1704_show_value(struct device *dev,
				  struct device_attribute *da, char *buf)
{
	struct sensor_device_attribute *attr = to_sensor_dev_attr(da);
	int value;
	int ret;

	ret = emc1704_get_value(dev_get_drvdata(dev), attr->index, &value);
	if (unlikely(ret < 0))
		return ret;

	return snprintf(buf, PAGE_SIZE, "%d\n", value);
}

static SENSOR_DEVICE_ATTR(internal_temp, S_IRUGO, emc1704_show_value, NULL,
			  EMC1704_INTERNAL_TEMP);
static SENSOR_DEVICE_ATTR(external1_temp, S_IRUGO, emc1704_show_value, NULL,
			  EMC1704_EXTERNAL1_TEMP);
static SENSOR_DEVICE_ATTR(external2_temp, S_IRUGO, emc1704_show_value, NULL,
				EMC1704_EXTERNAL2_TEMP);
static SENSOR_DEVICE_ATTR(external3_temp, S_IRUGO, emc1704_show_value, NULL,
				EMC1704_EXTERNAL3_TEMP);
static SENSOR_DEVICE_ATTR(source_voltage, S_IRUGO, emc1704_show_value, NULL,
			  EMC1704_SOURCE_VOLT_REGISTER);
static SENSOR_DEVICE_ATTR(sense_voltage, S_IRUGO, emc1704_show_value, NULL,
			  EMC1704_SENSE_VOLT_REGISTER);

static struct attribute *emc1704_attrs[] = {
	&sensor_dev_attr_internal_temp.dev_attr.attr,
	&sensor_dev_attr_external1_temp.dev_attr.attr,
	&sensor_dev_attr_external2_temp.dev_attr.attr,
	&sensor_dev_attr_external3_temp.dev_attr.attr,
	&sensor_dev_attr_source_voltage.dev_attr.attr,
	&sensor_dev_attr_sense_voltage.dev_attr.attr,
	NULL,
};
ATTRIBUTE_GROUPS(emc1704);

static int emc1704_i2c_probe(struct i2c_client *i2c,
			     const struct i2c_device_id *id)
{
	//int ret;
	struct device *hwmon_dev;

	if (!i2c_check_functionality(i2c->adapter, I2C_FUNC_SMBUS_BYTE_DATA |
				     I2C_FUNC_SMBUS_WORD_DATA))
		return -ENODEV;

	hwmon_dev = devm_hwmon_device_register_with_groups(&i2c->dev,
							   i2c->name,
							   i2c,
							   emc1704_groups);

	return PTR_ERR_OR_ZERO(hwmon_dev);
}

static const struct i2c_device_id emc1704_i2c_id[] = {
	{ "emc1704", 0 },
	{}
};
MODULE_DEVICE_TABLE(i2c, emc1704_i2c_id);

static struct i2c_driver emc17xx_i2c_driver = {
	.driver = {
		.name = "emc1704",
	},
	.probe    = emc1704_i2c_probe,
	.id_table = emc1704_i2c_id,
};

module_i2c_driver(emc17xx_i2c_driver);

MODULE_DESCRIPTION("EMC1704 Sensor Driver");
MODULE_AUTHOR("Mathew McBride <matt@traverse.com.au>");
MODULE_LICENSE("GPL v2");


emc2301/emc2301.c
/*
 * Driver for the Microchip/SMSC EMC2301 Fan controller
 *
 * Copyright (C) 2018-2020 Traverse Technologies
 * Author: Mathew McBride <matt@traverse.com.au>
 *
 * Based in part on an earlier implementation by
 * Reinhard Pfau <pfau@gdsys.de>
 *
 * SPDX-License-Identifier: GPL-2.0
 *
 */

#include <linux/err.h>
#include <linux/hwmon.h>
#include <linux/hwmon-sysfs.h>
#include <linux/i2c.h>
#include <linux/kernel.h>
#include <linux/module.h>
#include <linux/version.h>
#include <linux/thermal.h>

#if LINUX_VERSION_CODE <= KERNEL_VERSION(5, 2, 0)
#define HWMON_CHANNEL_INFO(stype, ...) \
	(&(struct hwmon_channel_info){ .type = hwmon_##stype, .config = (u32[]){ __VA_ARGS__, 0 } })
#endif
/*
 * Factor by equations [2] and [3] from data sheet; valid for fans where the
 * number of edges equals (poles * 2 + 1).
 */
#define FAN_RPM_FACTOR 3932160
#define FAN_TACH_MULTIPLIER 1

#define TACH_HIGH_MASK GENMASK(12, 5)
#define TACH_LOW_MASK GENMASK(4, 0)

#define EMC230X_REG_PRODUCT_ID 0xFD

#define EMC230X_REG_MINTACH 0x38

#define EMC230X_MAX_COOLING_STATE 7

static bool register_cdev = 1;
module_param(register_cdev, bool, 0);
MODULE_PARM_DESC(register_cdev, "Set to zero to disable registering a cooling device");

struct emc2301_data {
	struct device *dev;
	struct i2c_client *i2c;
	u8 num_fans;
	u16 minimum_rpm[5];
	struct thermal_cooling_device *cdev;

	u8 current_cooling_state;
	u16 cooling_step[5];
	u16 min_rpm[5];
	u16 max_rpm[5];
	u16 setpoint[5];
};

static u16 emc2301_read_fan_tach_int(struct emc2301_data *data, int channel)
{
	struct i2c_client *i2c = data->i2c;

	u8 channel_reg;
	u8 channel_high, channel_low;
	u16 channel_combined;

	channel_reg = 0x3e + (channel * 0x02);
#if 0
	dev_dbg(data->dev, "Reading channel %d register %X\n", channel, channel_reg);
#endif

	channel_high = i2c_smbus_read_byte_data(i2c, channel_reg);

	channel_low = i2c_smbus_read_byte_data(i2c, channel_reg + 0x01);
	channel_combined = ((u16)channel_high << 5) | (channel_low >> 3);

#if 0
	dev_dbg(data->dev, "Got values %04X for channel %d\n", channel_combined, channel);
#endif

	return channel_combined;
}

static u16 emc2301_read_fan_tach(struct device *dev, int channel)
{
	struct emc2301_data *data = dev_get_drvdata(dev);
	return emc2301_read_fan_tach_int(data, channel);
}

static int emc2301_read_fan_rpm(struct device *dev, int channel, long *val)
{
	u16 channel_tach;
	u16 fan_rpm;

	channel_tach = emc2301_read_fan_tach(dev, channel);

	fan_rpm = (FAN_RPM_FACTOR * FAN_TACH_MULTIPLIER) / channel_tach;
	*val = fan_rpm;

	return 0;
}

/* Write a target RPM to the TACH target register
 * This requires RPM speed control to be enabled with
 * EN_ALGO in the fan configuration register.
 */
static int emc2301_set_fan_rpm(struct emc2301_data *devdata, int channel, long target_rpm)
{
	long rpm_high, rpm_low;
	long target_tach;
	u8 channel_reg;

	channel_reg = 0x3c + (channel * 10);

	target_tach = (FAN_RPM_FACTOR * FAN_TACH_MULTIPLIER) / target_rpm;
	dev_dbg(devdata->dev, "RPM %ld requested, target tach is %ld\n", target_rpm, target_tach);

	rpm_high = (target_tach & TACH_HIGH_MASK) >> 5;
	rpm_low = (target_tach & TACH_LOW_MASK);

#if 0
	dev_dbg(devdata->dev, "Writing %02lX %02lX to %02X+%02X\n", rpm_low, rpm_high,
		channel_reg, channel_reg+1);
#endif

	i2c_smbus_write_byte_data(devdata->i2c, channel_reg, rpm_low);
	i2c_smbus_write_byte_data(devdata->i2c, channel_reg + 1, rpm_high);

	devdata->setpoint[channel] = (u16)target_rpm;

	return 0;
}

static int emc2301_fan_get_max_state(struct thermal_cooling_device *cdev, unsigned long *state)
{
	*state = EMC230X_MAX_COOLING_STATE;
	return 0;
}

static int emc2301_fan_get_cur_state(struct thermal_cooling_device *cdev, unsigned long *state)
{
	struct emc2301_data *devdata = cdev->devdata;
	*state = devdata->current_cooling_state;

	return 0;
}

static int emc2301_fan_set_cur_state(struct thermal_cooling_device *cdev, unsigned long state)
{
	struct emc2301_data *devdata = cdev->devdata;
	struct device *dev = devdata->dev;

	u16 rpm = 0;
	int i = 0;
	int retval = 0;

	dev_dbg(dev, "emc2301_fan_set_cur_state %ld\n", state);

	if (devdata->i2c == NULL) {
		dev_err(dev, "ERROR: no i2c instance\n");
		return -EINVAL;
	}

	if (state < 0 || state > EMC230X_MAX_COOLING_STATE)
		return -EINVAL;

	for (i = 0; i < ARRAY_SIZE(devdata->cooling_step); i++) {
		if (devdata->cooling_step[i] > 0) {
			rpm = devdata->min_rpm[i] + ((state + 1) * devdata->cooling_step[i]);
			devdata->current_cooling_state = state;
			if (emc2301_set_fan_rpm(devdata, i, rpm)) {
				retval = 1;
			}
		}
	}

	return retval;
}

struct thermal_cooling_device_ops emc2301_thermal_cooling_device = {
	.get_max_state = emc2301_fan_get_max_state,
	.get_cur_state = emc2301_fan_get_cur_state,
	.set_cur_state = emc2301_fan_set_cur_state,
};

static int emc2301_read_fan_fault(struct device *dev, struct i2c_client *i2c, int channel, long *val)
{
	u8 status_reg;

	if (channel > 1) {
		return -EOPNOTSUPP;
	}
	status_reg = i2c_smbus_read_byte_data(i2c, 0x24);
	dev_dbg(dev, "Channel %d status register %02X\n", channel, status_reg);

	if (status_reg & 0x7) {
		*val = 1;
	} else {
		*val = 0;
	}

	return 0;
}

static int emc2301_read_fan_target(struct emc2301_data *devdata, int channel, long *val)
{
	if (channel > devdata->num_fans) {
		return -EINVAL;
	}
	*val = devdata->setpoint[channel];
	return 0;
}

static int emc2301_read(struct device *dev, enum hwmon_sensor_types type, u32 attr, int channel, long *val)
{
	struct emc2301_data *data = dev_get_drvdata(dev);
	struct i2c_client *i2c = data->i2c;

	if (type != hwmon_fan) {
		return -EOPNOTSUPP;
	}
	if (channel > data->num_fans) {
		return -ENOTSUPP;
	}
	switch (attr) {
	case hwmon_fan_input:
		return emc2301_read_fan_rpm(dev, channel, val);
	case hwmon_fan_target:
		return emc2301_read_fan_target(data, channel, val);
	case hwmon_fan_fault:
		return emc2301_read_fan_fault(dev, i2c, channel, val);
	default:
		return -ENOTSUPP;
		break;
	}

	return 0;
}

static int emc2301_write(struct device *dev, enum hwmon_sensor_types type, u32 attr, int channel, long val)
{
	struct emc2301_data *data = dev_get_drvdata(dev);

	if (channel > data->num_fans)
		return -EINVAL;

	switch (type) {
	case hwmon_fan:
		switch (attr) {
		case hwmon_fan_target:
#if 0
			if (val < data->minimum_rpm[channel]) {
				dev_err(dev, "RPM %ld is lower than channel minimum %ld\n", val, data->minimum_rpm[channel]);
				return -EINVAL;
			}
#endif

			dev_dbg(dev, "emc2301_write hwmon_fan_target %ld\n", val);
			return emc2301_set_fan_rpm(data, channel, val);
		default:
			return -EOPNOTSUPP;
		}
	default:
		return -EOPNOTSUPP;
		break;
	}
	return -EOPNOTSUPP;
}

static const struct hwmon_channel_info *emc2301_info[] = {
	HWMON_CHANNEL_INFO(fan, HWMON_F_INPUT | HWMON_F_FAULT | HWMON_F_TARGET), NULL
};

static umode_t emc2301_is_visible(const void *drvdata, enum hwmon_sensor_types type, u32 attr, int channel)
{
	//const struct emc2301_data *data = drvdata;

	switch (type) {
	case hwmon_fan:
		switch (attr) {
		case hwmon_fan_input:
		case hwmon_fan_fault:
			return S_IRUGO;
		case hwmon_fan_target:
			return S_IRUGO | S_IWUSR;
		default:
			break;
		}
		break;
	default:
		break;
	}
	return 0;
}

static const struct hwmon_ops emc2301_ops = {
	.is_visible = emc2301_is_visible,
	.read = emc2301_read,
	.write = emc2301_write,
};

static const struct hwmon_chip_info emc2301_chip_info = {
	.ops = &emc2301_ops,
	.info = emc2301_info,
};

static int emc2301_enable_rpm_control(struct emc2301_data *data, u16 fan_dev, bool enable)
{
	u8 fan_config_reg_addr;
	u8 fan_config_reg_val;
	int ret = 0;

	// get current fan config reg value
	fan_config_reg_addr = 0x32 + (fan_dev * 0x10);
	fan_config_reg_val = i2c_smbus_read_byte_data(data->i2c, fan_config_reg_addr);

	// update config reg to enable/disable control as requested
	if (enable) {
		// set ENAx to enable drive
		fan_config_reg_val |= (1 << 7);
		// clear RNGx to set minRPM=500
		fan_config_reg_val &= ~(0b11 << 5);
	} else {
		// clear ENAx
		fan_config_reg_val &= ~(1 << 7);
	}

	dev_dbg(data->dev, "Writing 0x%02X to 0x%02X\n", fan_config_reg_val, fan_config_reg_addr);

	ret = i2c_smbus_write_byte_data(data->i2c, fan_config_reg_addr, fan_config_reg_val);
	if (ret) {
		dev_err(data->dev, "Unable to write fan configuration register %02X\n", fan_config_reg_addr);
		return ret;
	}

	if (!enable) {
		ret = i2c_smbus_write_byte_data(data->i2c, (0x30 + (fan_dev * 0x10)), 0xFF);
	}
	return ret;
};

static int emc2301_i2c_probe(struct i2c_client *i2c, const struct i2c_device_id *id)
{
	struct device *hwmon_dev;
	struct device_node *of_node = i2c->dev.of_node;
	struct device_node *child_node;
	struct emc2301_data *data;
	int8_t regval;
	u8 i, retval;
	u16 chan_val;
	u16 range;
	bool has_cooling_step = false;
	int numchildren = 0;

	if (!i2c_check_functionality(i2c->adapter, I2C_FUNC_SMBUS_BYTE_DATA | I2C_FUNC_SMBUS_WORD_DATA))
		return -ENODEV;

	data = devm_kzalloc(&i2c->dev, sizeof(struct emc2301_data), GFP_KERNEL);
	if (unlikely(!data))
		return -ENODEV;

	data->dev = &i2c->dev;
	data->i2c = i2c;

	regval = i2c_smbus_read_byte_data(i2c, EMC230X_REG_PRODUCT_ID);
	switch (regval) {
	case 0x34: /* EMC2305 */
		data->num_fans = 5;
		break;
	case 0x35: /* EMC2303 */
		data->num_fans = 3;
		break;
	case 0x36: /* EMC2302 */
		data->num_fans = 2;
		break;
	case 0x37: /* EMC2301 */
		data->num_fans = 1;
		break;
	default:
		dev_err(&i2c->dev, "Unknown product ID %d\n", regval);
		return -ENODEV;
	}
	dev_info(&i2c->dev, "EMC230%d detected\n", data->num_fans);

	memset(data->min_rpm, 0, sizeof(u16) * ARRAY_SIZE(data->min_rpm));
	memset(data->max_rpm, 0, sizeof(u16) * ARRAY_SIZE(data->max_rpm));

	/* Read minimum and maximum RPM values from device
	 * tree, if specified.
	 * For example, the Noctua NF-A4x20 has a min
	 * RPM of 1200 and a max of 5000-5500.
	 * Without this, cooling actions with slower
	 * fans may not be effective
	 */
	if (of_node) {
		numchildren = of_get_child_count(of_node);
		dev_info(&i2c->dev, "Have %d fans configured in DT\n", numchildren);
		if (numchildren > data->num_fans) {
			dev_warn(&i2c->dev,
				 "%d fans are specified in DT, but only %d fans are supported by this device",
				 numchildren, data->num_fans);
			numchildren = data->num_fans;
		}
		child_node = of_node->child;

		for (i = 0; i < numchildren; i++) {
			retval = of_property_read_u16(child_node, "min-rpm", &chan_val);
			if (!retval) {
				data->min_rpm[i] = chan_val;
			}

			retval = of_property_read_u16(child_node, "max-rpm", &chan_val);
			if (!retval) {
				data->max_rpm[i] = chan_val;
			}
			child_node = of_get_next_child(of_node, child_node);
		}
	} else {
		dev_warn(&i2c->dev, "No device tree node found for this device");
	}

	for (i = 0; i < data->num_fans; i++) {
		if (data->min_rpm[i] != 0 && data->max_rpm[i] != 0) {
			range = data->max_rpm[i] - data->min_rpm[i];
			data->cooling_step[i] = range / (EMC230X_MAX_COOLING_STATE + 1);
			dev_info(&i2c->dev, "Fan %i Cooling step is %d RPM, minimum %d, max %d RPM\n", i,
				 data->cooling_step[i], data->min_rpm[i], data->max_rpm[i]);
			has_cooling_step = true;
			emc2301_enable_rpm_control(data, i, true);
			emc2301_set_fan_rpm(data, i, data->max_rpm[i]);
		} else {
			data->cooling_step[i] = 0;
		}
	}

	data->current_cooling_state = EMC230X_MAX_COOLING_STATE;

#if 0
	/* Read the fan minimum tach values */
	for(i=0; i<data->num_fans; i++) {
		channel_reg = EMC230X_REG_MINTACH + (i * 10);
		regval = i2c_smbus_read_byte_data(i2c, channel_reg);
		min_tach = (FAN_RPM_FACTOR * FAN_TACH_MULTIPLIER) / (regval << 5);
		data->minimum_rpm[i] = (FAN_RPM_FACTOR * FAN_TACH_MULTIPLIER) / min_tach;
		dev_info(&i2c->dev, "Channel %d minimum RPM is %d", i, data->minimum_rpm[i]);
	}
#endif
	hwmon_dev =
		devm_hwmon_device_register_with_info(&i2c->dev, i2c->name, data, &emc2301_chip_info, NULL);

	if (IS_REACHABLE(CONFIG_THERMAL) && has_cooling_step && register_cdev == 1) {
		dev_info(&i2c->dev, "registering a cooling device");
		data->cdev = devm_thermal_of_cooling_device_register(&i2c->dev, of_node, "emc2301_fan", data,
								     &emc2301_thermal_cooling_device);
		if (IS_ERR(data->cdev)) {
			dev_err(&i2c->dev, "Failed to register cooling device\n");
			return PTR_ERR(data->cdev);
		}
	}

	return PTR_ERR_OR_ZERO(hwmon_dev);
}

static const struct i2c_device_id emc2301_i2c_id[] = { { "emc2305", 0 }, { "emc2304", 0 }, { "emc2303", 0 },
						       { "emc2302", 0 }, { "emc2301", 0 }, {} };

MODULE_DEVICE_TABLE(i2c, emc2301_i2c_id);

static struct i2c_driver emc2301_i2c_driver = {
	.driver = {
		.name = "emc2301",
	},
	.probe    = emc2301_i2c_probe,
	.id_table = emc2301_i2c_id,
};

module_i2c_driver(emc2301_i2c_driver);

MODULE_DESCRIPTION("EMC2301 Fan controller driver");
MODULE_AUTHOR("Mathew McBride <matt@traverse.com.au>");
MODULE_LICENSE("GPL v2");
